### ISSUE TITLE
No default TypeMap specified

### ISSUE THREAD

**Issue by nicain** [NeurodataWithoutBorders/pynwb#114]
when running a little hellow world writer:

https://gist.github.com/nicain/ae3d5280d86843971db8bd6dd0137948

And then in a separate script trying to read the file:

```
from form.backends.hdf5 import HDF5IO
io2 = HDF5IO('tmp.nwb', mode='r')
print(io2)
```

I get:  

```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-2-02fcd47d665e> in <module>()
----> 1 io2 = HDF5IO('tmp.nwb', mode='r')

/home/nicholasc/projects/pynwb/src/form/utils.pyc in func_call(*args, **kwargs)
    270                     msg = ', '.join(parse_err)
    271                     raise_from(TypeError(msg), None)
--> 272                 return func(self, **parsed['args'])
    273         else:
    274             def func_call(*args, **kwargs):

/home/nicholasc/projects/pynwb/src/form/backends/hdf5/h5tools.pyc in __init__(self, **kwargs)
     23         '''
     24         path, manager, mode = popargs('path', 'manager', 'mode', kwargs)
---> 25         super(HDF5IO, self).__init__(manager, source=path)
     26         self.__path = path
     27         self.__mode = mode

/home/nicholasc/projects/pynwb/src/form/utils.pyc in func_call(*args, **kwargs)
    270                     msg = ', '.join(parse_err)
    271                     raise_from(TypeError(msg), None)
--> 272                 return func(self, **parsed['args'])
    273         else:
    274             def func_call(*args, **kwargs):

/home/nicholasc/projects/pynwb/src/form/backends/io.pyc in __init__(self, **kwargs)
     14             type_map = TypeMap.default()
     15             if type_map is None:
---> 16                 raise RuntimeError('No default TypeMap specified')
     17             self.__manager = BuildManager(type_map)
     18         self.__built = dict()

RuntimeError: No default TypeMap specified
```

**Comment by nicain** [comment]
When I run:
from pynwb import NWBFile
from form.backends.hdf5 import HDF5IO
io2 = HDF5IO('tmp.nwb', mode='r')

There are no problems

**Comment by ajtritt** [comment]
This isn't really a bug. It's part of the current design of the system. Without loading _pynwb_, _form_ has no idea what it is supposed to read. Loading _pynwb_ registers the NWB specs with _form_. Until we have the specifications dumped to file during write (see #44), you will have to load _pynwb_ to register the NWB specs with _form_.

**Comment by neuromusic** [comment]
in that case, might need to make the error more clear

feel free to adjust the milestone as you see necessary @ajtritt

**Comment by neuromusic** [comment]
thinking about this more... 
"Loading pynwb registers the NWB specs with form."

This happens behind the scenes? via the pynwb import setting some global variables or something, I presume?

In general, that is going to be super confusing for users, as it was for @nicain. PEP 20: "Explicit is better than implicit."

While #44 is a good solution, it does not address the case where one might want to load an NWB file written with the basic specification using a feature-rich specification. 

I would recommend that HDF5IO take an optional "specification" argument or something that explicitly tells HDF5IO which specification to use. If it is not passed, HDF5IO relies on #44. If it is passed, it overrides the file's self-describing specification.

**Comment by oruebel** [comment]
Maybe one way to satisfy all of the above concerns (i.e., 1. being explicit, 2. having FORM agnostic to the format, and 3. not requiring users to specify the schema when using NWB) we could maybe have a very light-weight I/O package in PyNWB itself which would essentially do nothing else but load the  I/O backend(s) from FORM and set the schema. In this way a user would also not need to go to FORM directly but stay in PyNWB which may make things at least look a bit simpler. Just an idea.

**Comment by neuromusic** [comment]
I think that's a great idea.

**Comment by ajtritt** [comment]
Can someone provide an example what you would want the user to do?

**Comment by neuromusic** [comment]
``` Python
# tell HDF5IO what spec to use
from my_package.specifications import bells_and_whistles
from form.backends.hdf5 import HDF5IO
io2 = HDF5IO('tmp.nwb', mode='r', specification=bells_and_whistles)
```

``` Python
# discover spec from `tmp.nwb` file
from form.backends.hdf5 import HDF5IO
io2 = HDF5IO('tmp.nwb', mode='r')
```

**Comment by ajtritt** [comment]
what is `bells_and_whistles`?

I should also note that this feature would allow only the use of dynamically created classes.

**Comment by neuromusic** [comment]
A specification which has more bells and whistles than the specification which the file was created with, but which the file would still validate against.

**Comment by ajtritt** [comment]
By 'bells and whistles' do you mean extensions? Do you mean user defined classes that extend the existing pynwb classes?

**Comment by nicain** [comment]
> we could maybe have a very light-weight I/O package in PyNWB itself which would essentially do nothing else but load the I/O backend(s) from FORM and set the schema.

Exactly!  In general, having too much going on behind the scenes is bad design. Consider the case of a user loading two files, with two different specs, in the same python instance.  Holding any variables at global level makes this really confusing; how does the loader know which spec to use if it is defined at global scope? 

Better to have a lightweight read object (created by a lightweight module) contain the spec for loading, which can safely be balled up and thrown away if a different loading spec is needed in the same python exectution session.

An alternative (complimentary?), and potentially very elegant solution, could use python context managers ("with" statement):

```Python
from form.backends.hdf5 import HDF5IO
from pynwb.somewhere import nwbschema
with nwbschema as specification:
    io2 = HDF5IO('tmp.nwb', mode='r')
```

**Comment by oruebel** [comment]
Ok, what I was thinking of was the following:

```from pynwb.io import HDFIO```

where in pynwb.io would essentially do nothing else but to import FORM and set the spec that FORM should use?

**Comment by neuromusic** [comment]
@ajtritt 
> By 'bells and whistles' do you mean extensions? Do you mean user defined classes that extend the existing pynwb classes?"

yes. perhaps a better example is the inverse (or is it the converse?)

Let's say I have a file someone gave me that was created with an extended specification. We'll call this extended specification `bells_and_whistles` ... it has some extensions... perhaps it has some attributes that are marked as "required" that are optional is the core spec?

pynwb should handle this easily by default

``` Python
# discover spec from `bells_and_whistles.nwb` file
from pynwb.io import HDF5IO
io2 = HDF5IO('bells_and_whistles.nwb', 'r')
```

However, let's say I have an analysis package or visualization package that works only on the core specification. I should be able to explicitly tell HDF5IO which specification I want to use.

``` Python
# open `bells_and_whistles.nwb` file with core specification
from pynwb.io import HDF5IO
from pynwb.specifications import core
io2 = HDF5IO('bells_and_whistles.nwb', 'r', specification=core)
```

Maybe my analysis package doesn't actually want to use the core specification, but rather sits "in between" `core` and `bells_and_whistles`... e.g. it wants to enforce certain required attributes, but is still unaware of the extensions in `bells_and_whistles`

``` Python
# open `bells_and_whistles.nwb` file with 'just_the_bells` specification
from pynwb.io import HDF5IO
from my_package.specifications import just_the_bells
io2 = HDF5IO('bells_and_whistles.nwb', 'r', specification=just_the_bells)
```

**Comment by neuromusic** [comment]
@oruebel 
> where in pynwb.io would essentially do nothing else but to import FORM 

yes.

> and set the spec that FORM should use?

no. this should absolutely not be set globally on import, for precisely the reasons that @nicain mentioned and that my examples highlight.

if it is not discovered from the file, then it should be explicitly given by the user.

**Comment by neuromusic** [comment]
@oruebel unless you mean that the pynwb.io.HDF5IO class uses the core pynwb spec as default? I'm OK with that.

**Comment by nicain** [comment]
I ran into this type of issue working with the Brian simulator (this was several years ago, so this may have changed).  I would set some populations, settings, etc to be simulated and the simulator would auto-discover these things, throw them into the simulation time-loop, and run, which on its face seemed very nice.  Except when it came time to average the results of multiple simulations.  Instead of just running a loop, the architecture required that use a different python session for each simulation, and save the results to disk.  Now I have to write a bash script to drive the loop,  saver/loader pair, discover the file names, and write a separate analysis file.  Not fun.  Ever since that experience, my rule is to never use globals for anything.  Especially when classes are so easy to write, and context management is built into the language.

The analogy here is loading multiple nwb files with slightly different extensions (i.e. customization from different labs).

**Comment by nicain** [comment]
@neuromusic 

> @oruebel unless you mean that the HDF5IO class uses the core pynwb spec as default? I'm OK with that.

HDF5IO is a part of form. This would introduce a coupling between form and pynwb... not sure it that is the intended architecture.

**Comment by neuromusic** [comment]
@nicain I corrected my comment: "the pynwb.io.HDF5IO class"

``` Python
from pynwb.io import HDF5IO # <- this class uses the pynwb core spec as default, 
#but doesn't do anything to the global namespace

# ... do some things

# now, I want to use FORM to load something totally different and I can do this b/c 
# there is no global spec set
from form.backends.hdf5 import HDF5IO #<- this one does not have a default spec 
# and can only autodiscover from the file or take an explicit specification from the user
```

**Comment by ajtritt** [comment]
Does this solution work? To be clear, the extensions in the namespaces are ANY extension of the core namespace. The only requirement here is that they are extensions of the CORE namespace.

``` python
import pynwb

foo_ext_path = 'foo.namespace.yaml' # custom extension
foo_nwb_path = 'foo.nwb'            # file using custom extension
bar_ext_path = 'bar.namespace.yaml' # another custom extension
bar_nwb_path = 'bar.nwb'            # file using this other custom extension
baz_nwb_path = 'baz.nwb'            # file with no extensions

foo_nwb = pynwb.read_hdf5(foo_nwb_path, extensions=foo_ext_path)
bar_nwb = pynwb.read_hdf5(bar_nwb_path, extensions=bar_ext_path)
baz_nwb = pynwb.read_hdf5(baz_nwb_path)
```

We could do something similar with write, but I suspect writing datasets with two different extensions at the same time is less relevant of a use-case.

**Comment by neuromusic** [comment]
yeah, though I also might want to do this...

``` Python
allen_nwb = pynwb.read_hdf5(
    frank_nwb_path, 
    extensions=allen_ext_path, # <- see what I did there?
)
```

> The only requirement here is that they are extensions of the CORE namespace.

I'm still wrapping my head around the namespace framework. What if I want to open a 2.0 file using the 2.1 specification? Or visa versa? How would that work?

**Comment by ajtritt** [comment]
> yeah, though I also might want to do this...

Can you give some explanation as to why you would want to do that? I ask becase, to me, that is the equivalent of saying "read from database X, but ignore X's schema, and use database Y's schema instead"

**Comment by oruebel** [comment]
I don't know if this makes sense. If you write data with a given schema then it should always be read with that schema. The only use-case I could see is that when you want to convert data from schema A to schema B, but I think that use-case needs to be handled differently. In particular, how do you know which fields in schema A map to schema B and vice-versa without having additional information about how the two schema map to each other?

**Comment by neuromusic** [comment]
> If you write data with a given schema then it should always be read with that schema.

not necessarily. let's say I have a visualization tool built for NWB 2.1 and I have an NWB 2.0 file that I want to load.  let's also say that the differences between the two schemas are in a module that the file may or may not implement.

the pythonic way to do this (EAFP) is to tell pynwb to open the file w/ the 2.1 schema regardless of what the file says it's schema is. if it works, it works. if it doesn't, it fails and the user has to decide what to do. 

``` Python
try:
    nwb = pynwb.read_hdf5(
        'nwb_two_point_oh.nwb',
        schema='nwb.2.1',
    )
except IOError:
    print('woops! that didn't work. better try something else.')
```

**Comment by ajtritt** [comment]
With the way we've designed extensions in NWB, anything written as NWB 2.0 should be compatible with a NWB 2.1 specification. You are not allowed to remove anything from parent types i.e. you are not allowed to modify or replace types, only create additional, more-specific specifications of existing types through means of creating a new type. Any type that was used to create nwb_two_point_oh.nwb should exist in the nwb.2.1 schema, so this call wouldn't raise an error.

**Comment by ajtritt** [comment]
I wrote up the I/O scenarios I could think of to help guide fixing this issue. I added some snippets (all in one script) demonstrating how read, write, and append scenarios would work with multiple extensions using the solution I have in mind.

https://gist.github.com/ajtritt/11f2bda4846a96eb67cd9e9a7e1b2bdc

Let me know if you think there are any that are missed any scenarios or limitations (by adding comments to the Gist, preferably).
